crates/bifrost/src/providers/replicated_loglet/sequencer/node.rs

```diff
@@ -0,0 +1,236 @@
+pub(crate) struct LogServerManager {
+    servers: Arc<BTreeMap<PlainNodeId, LogServerLock>>,
+    node_set: NodeSet,
+    metadata: Metadata,
+    connection_manager: ConnectionManager,
+}
+
+impl Clone for LogServerManager {
+    fn clone(&self) -> Self {
+        Self {
+            servers: Arc::clone(&self.servers),
+            node_set: self.node_set.clone(),
+            metadata: self.metadata.clone(),
+            connection_manager: self.connection_manager.clone(),
+        }
+    }
+}
+
+impl LogServerManager {
+    /// creates the node set and start the appenders
+    pub fn new(
+        metadata: Metadata,
+        connection_manager: ConnectionManager,
+        node_set: NodeSet,
+    ) -> Result<Self, super::Error> {
+        let mut servers = BTreeMap::default();
+        for node_id in node_set.iter() {
+            servers.insert(*node_id, LogServerLock::default());
+        }
+
+        Ok(Self {
+            servers: Arc::new(servers),
+            node_set,
+            metadata,
+            connection_manager,
+        })
+    }
+
+    async fn connect(&self, id: PlainNodeId) -> Result<ConnectionSender, Error> {
+        let conf = self.metadata.nodes_config_ref();
+        let node = conf.find_node_by_id(id)?;
+        let connection = self
+            .connection_manager
+            .get_node_sender(node.current_generation)
+            .await?;
+
+        Ok(connection)
+    }
+
+    /// gets a log-server instance. On fist time it will initialize a new connection
+    /// to log server. It will make sure all following get call will hold the same
+    /// connection.
+    ///
+    /// it's up to the client to call [`Self::renew`] if the connection it holds
+    /// is closed
+    pub async fn get(&self, id: PlainNodeId) -> Result<LogServer, Error> {
+        let server = self.servers.get(&id).ok_or(Error::InvalidNodeSet)?;
+
+        let mut guard = server.0.lock().await;
+
+        if let Some(current) = guard.deref() {
+            return Ok(current.clone());
+        }
+
+        // initialize a new instance
+        let server = LogServer {
+            node: id,
+            state: LogServerState::default(),
+            sender: self.connect(id).await?,
+        };
+
+        // we need to update initialize it
+        *guard = Some(server.clone());
+
+        Ok(server)
+    }
+
+    /// renew makes sure server connection is renewed if and only if
+    /// the provided server holds an outdated connection. Otherwise
+    /// the latest connection associated with this server is used.
+    ///
+    /// It's up the holder of the log server instance to retry to renew
+    /// if that connection is not valid.
+    ///
+    /// It also grantees that concurrent call to renew on the same server instance
+    /// will only renew the connection once for all callers
+    pub async fn renew(&self, server: &mut LogServer) -> Result<(), Error> {
```

Comment:
I wonder how will this be called from appenders, are we going to share a reference to LogServerManager with all appenders? Wouldn't we have lifetime constraint that it needs to be a `'static` reference?